tests/bsim/bluetooth/host/att/timeout/main.c

```diff
@@ -0,0 +1,292 @@
+{
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_ERR);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_ERR);
+
+	EXPECT_ZERO(bt_enable(NULL));
+
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_INF);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_INF);
+}
+
+static struct bt_conn *peripheral_setup(enum bt_att_chan_opt bearer, bool timeout)
+{
+	struct bt_conn *conn = NULL;
+
+	EXPECT_ZERO(bt_testlib_adv_conn(&conn, BT_ID_DEFAULT, bt_get_name()));
+
+	trigger_att_timeout = timeout;
+
+	return conn;
+}
+
+static struct bt_conn *central_setup(enum bt_att_chan_opt bearer, bool timeout)
+{
+	bt_addr_le_t adva;
+	struct bt_conn *conn = NULL;
+
+	EXPECT_ZERO(bt_testlib_scan_find_name(&adva, "peripheral"));
+	EXPECT_ZERO(bt_testlib_connect(&adva, &conn));
+
+	/* Establish EATT bearers. */
+	EXPECT_ZERO(bt_testlib_secure(conn, BT_SECURITY_L2));
+
+	while (bt_eatt_count(conn) == 0) {
+		k_msleep(100);
+	};
+
+	return conn;
+}
+
+static void central_read(struct bt_conn *conn, enum bt_att_chan_opt bearer, bool timeout)
+{
+	uint16_t actual_read_len;
+	uint16_t remote_read_send_len;
+	uint16_t handle = 0;
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(attr_value, sizeof(remote_read_send_len));
+
+	err = bt_testlib_att_read_by_type_sync(&attr_value, &actual_read_len, &handle, NULL, conn,
+					       bearer, UUID_2, BT_ATT_FIRST_ATTRIBUTE_HANDLE,
+					       BT_ATT_LAST_ATTRIBUTE_HANDLE);
+
+	if (timeout) {
+		__ASSERT(err == BT_ATT_ERR_UNLIKELY, "Unexpected error %d", err);
+	} else {
+		__ASSERT(!err, "Unexpected error %d", err);
+		__ASSERT(attr_value.len >= sizeof(remote_read_send_len),
+			 "Remote sent too little data.");
+		remote_read_send_len = net_buf_simple_pull_le16(&attr_value);
+		__ASSERT(remote_read_send_len == actual_read_len, "Length mismatch. %u %u",
+			 remote_read_send_len, actual_read_len);
+	}
+}
+
+/**
+ * Test procedure:
+ *
+ * Central:
+ * 1. Connect to the peripheral.
+ * 2. Try to read a characteristic value.
+ * 3. Expect BT_ATT_ERR_UNLIKELY error.
+ * 4. Expect the peripheral to disconnect.
+ * 5. Reconnect to the peripheral.
+ * 6. Try to read a characteristic value.
+ * 7. Expect the peripheral to respond with the characteristic value.
+ * 8. Ensure that connection stays alive after a delay equal to ATT timeout.
+ * 9. Disconnect from the peripheral.
+ *
+ * Peripheral:
+ * 1. Start advertising.
+ * 2. Make the read callback sleep for more than ATT Timeout when the central tries to read.
+ * 3. Expect the disconnected callback to be called.
+ * 4. Start advertising again.
+ * 5. Make the read callback respond with the characteristic value when the central tries to read.
+ * 6. Expect the connection stay alive after a delay equal to ATT timeout.
+ * 7. Expect the central to disconnect.
+ */
+static void test_timeout(enum bt_att_chan_opt bearer)
+{
+	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
+	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);
+	struct bt_conn *conn;
+	int err;
```

Comment:
Function is much better now, thanks :)

My personal preference would still be to have 2 separate functions (one for central, one for peripheral), since they share so little code, but this is OK